baseclasses/utils/solverHistory.py

```diff
@@ -0,0 +1,354 @@
+        data : Any
+            Item to store
+        """
+        self.metadata[name] = data
+
+    def addVariable(
+        self,
+        name: str,
+        varType: Type,
+        printVar: bool = False,
+        printFormat: Optional[str] = None,
+    ) -> None:
+        """Define a new field to be stored in the history.
+
+        Parameters
+        ----------
+        name : str
+            Variable name
+        varType : Type
+            Variable type, i.e int, float, str etc
+        printVar : bool, optional
+            Whether to include the variable in the iteration printout, by default False
+        printFormat : str, optional
+            Format string valid for use with the str.format() method (e.g "{:17.11e}" for a float or "{:03d}" for an
+            int), by default None, in which case a default format for the given `varType` is used, not used if
+            `printVar` is False
+        """
+        self.variables[name] = {"print": printVar, "type": varType}
+
+        # Create variable data list
+        self.data[name] = []
+
+        # Only store variable's string format if it's going to be printed
+        if printVar:
+            if printFormat is not None:
+                self.variables[name]["format"] = printFormat
+            elif varType in self.defaultFormat:
+                self.variables[name]["format"] = self.defaultFormat[varType]
+            else:
+                self.variables[name]["format"] = self.DEFAULT_OTHER_FORMAT
+
+            # Get test value for figuring out how long a string is using the supplied format
+            try:
+                testValue = self.testValues[varType]
+            except KeyError:
+                testValue = self.DEFAULT_OTHER_VALUE
+
+            # Figure out column width, the maximum of the length of the name and the formatted value, also check that
+            # the format string is valid
+            try:
+                testString = self.variables[name]["format"].format(testValue)
+            except (ValueError, TypeError) as e:
+                raise ValueError(f"Supplied format string \"{self.variables[name]['format']}\" is invalid") from e
+
+            dataLen = len(testString)
+            nameLen = len(name)
+            self.variables[name]["columnWidth"] = max(dataLen, nameLen)
+
+            # --- Now figure out the format strings for the iteration printout header and values ---
+            # The header is simply centred in the available width, which is actually columnWidth + 4
+            self.variables[name]["headerFormat"] = f"{{:^{(self.variables[name]['columnWidth'] + 4)}s}}"
+
+    def startTiming(self) -> None:
+        """Record the start time of the solver
+
+        This function only needs to be called explicitly if the start time of your solver is separate from the first
+        time the `write` method is called.
+        """
+        self.startTime = time.time()
+
+    def write(self, data: dict) -> None:
```

Comment:
It would be good to have another function that writes data for only one variable. Here you have to collect everything into one dict to then provide this function.

Reply:
I'm not sure about this. The choice to have the user write all variables was a deliberate one on my part. I particularly wanted to support certain variables that might not have a value at every iteration, in which case you just exclude them from the call to `write` and a value of `None` is appended to their history for that iteration.

If we had a method for recording variables one by one then these variables would get out of sync with those that have a value every iteration.

Reply:
But there is only one iteration counter, so as soon as you allow writing one variable at a time you have no guarantee that any of the variables have a consistent number of entries with each other or with the iteration counter. Unless there is a really compelling example where you absolutely cannot collect all the variables together and write them in one call then I would like to keep it like this.